src/osd/PrimaryLogPG.h

```diff
@@ -1622,18 +1622,25 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
       > reactions;
     struct ReservationCB : public Context {
       PrimaryLogPGRef pg;
+      OSDService *service;
       bool canceled;
-      ReservationCB(PrimaryLogPG *pg) : pg(pg), canceled(false) {}
+      ReservationCB(PrimaryLogPG *pg) : pg(pg), service(pg->osd),
+					canceled(false) {}
       void finish(int) override {
-	pg->lock();
+	PrimaryLogPG *pgp = pg.get();
+	if (service->is_stopping() || !pgp)
+	  return;
+	pgp->lock();
 	if (!canceled)
-	  pg->snap_trimmer_machine.process_event(SnapTrimReserved());
-	pg->unlock();
+	  pgp->snap_trimmer_machine.process_event(SnapTrimReserved());
```

Comment:
We need to unlock and there aren't any Locker semantics around the extra PG lock interfaces that I'm aware of (not that it would be any shorter in this case).